### PR TITLE
[scheduler] hide StringIDMap under BaseSchedulingID

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

Today, we use StringIdMap to optimize access to node_id and resource_name, which stores the mapping from string to int ids. 
However, the implementation passes the StringIdMap around and makes the code bloated and hard to read.

This PRs tries to solve the problem but retain this optimization, by introducing BaseSchedulingID type to hide both the StringIdMap and implementation details. Specifically, the BaseSchedulingID type wraps around the id_ integer, and shares a singleton StringIdMap.

The major change is in scheduling_ids.h

Why don't we get rid of this optimization? Turns out the scheduling_policy does a lot operations on node_id and slows down the performance if we use string instead of int for node_id.  (https://github.com/ray-project/ray/pull/22662 has more details)

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/scheduling/scheduling_ids.h [hunk 1]
```diff
@@ -62,3 +74,78 @@ class StringIdMap {
...
+  bool operator<(const BaseSchedulingID &rhs) const { return id_ < rhs.id_; }
+
+  bool IsNull() const { return id_ == -1; }
+
+  static BaseSchedulingID NullID() { return BaseSchedulingID(-1); }
+
+ private:
+  /// Meyer's singleton to store the StringIdMap.
+  static StringIdMap &GetMap() {
+    static StringIdMap map;
```
[the remark this answers is not in the document]
Comment (scv119) [reply src/ray/raylet/scheduling/scheduling_ids.h:109]: it'll be as good(bad) as today...

File: src/ray/raylet/scheduling/scheduling_ids_test.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  for (auto &string_id : string_ids) {
+    node_ids.emplace_back(NodeID(string_id));
+    ASSERT_EQ(node_ids.back().ToBinary(), string_id);
+  }
+  ASSERT_EQ(node_ids[0], NodeID(string_ids[0]));
+  ASSERT_EQ(node_ids[0], NodeID(node_ids[0].ToInt64()));
+
+  ASSERT_TRUE(NodeID::NullID().IsNull());
+  ASSERT_EQ(NodeID::NullID().ToInt64(), -1);
+  ASSERT_EQ(NodeID::NullID().ToBinary(), "-1");
```
[the remark this answers is not in the document]
Comment (scv119) [reply src/ray/raylet/scheduling/scheduling_ids_test.cc]: we don't support +/-, but added other tests!

**Review by WangTaoTheTonic — APPROVED** [review]
This is very nice! Glad to see the `string_to_int_map` being under control.

File: src/ray/raylet/scheduling/scheduling_ids.h [hunk 3]
```diff
@@ -62,3 +74,85 @@ class StringIdMap {
...
+template <SchedulingIDTag T>
+class BaseSchedulingID {
+ public:
+  explicit BaseSchedulingID(const std::string &name) : id_{GetMap().Insert(name)} {}
+
+  explicit BaseSchedulingID(int64_t id) : id_{id} {}
+
+  int64_t ToInt() const { return id_; }
+
+  std::string Binary() const { return GetMap().Get(id_); }
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply src/ray/raylet/scheduling/scheduling_ids.h:94]: I think `Binary()` is correct here (for other id's we also have `Hex()` which returns it as a string)

**Comment by scv119** [comment]
master benchmark:
```
{
   "single_client_get_calls_Plasma_Store":[
      5882.07192336896,
      163.53542379085826
   ],
   "single_client_put_calls_Plasma_Store":[
      4427.858679780679,
      14.56568724293238
   ],
   "multi_client_put_calls_Plasma_Store":[
      9637.362199422822,
      47.85051433630265
   ],
   "single_client_put_gigabytes":[
      15.657111235158766,
      3.283226569931537
   ],
   "single_client_tasks_and_get_batch":[
      9.850286152675713,
      0.11388791642914735
   ],
   "multi_client_put_gigabytes":[
      6.634510320471843,
      0.5647390431984716
   ],
   "single_client_get_object_containing_10k_refs":[
      14.976534587453802,
      0.02246055444501522
   ],
   "single_client_tasks_sync":[
      1522.9645470222172,
      9.073900830808936
   ],
   "single_client_tasks_async":[
      9564.642146240098,
      74.50053980655738
   ],
   "multi_client_tasks_async":[
      10746.516886472014,
      175.98873564507508
   ],
   "1_1_actor_calls_sync":[
      2617.04455735718,
      19.029600988044734
   ],
   "1_1_actor_calls_async":[
      6034.114081282305,
      102.28477950668156
   ],
   "1_1_actor_calls_concurrent":[
      4964.132462657995,
      53.08419203372572
   ],
   "1_n_actor_calls_async":[
      9455.182457364925,
      90.55394985303715
   ],
   "n_n_actor_calls_async":[
      12944.873939688503,
      522.3001858049639
   ],
   "n_n_actor_calls_with_arg_async":[
      3233.077733868373,
      81.90205123907317
   ],
   "1_1_async_actor_calls_sync":[
      1785.334128044559,
      12.936645699204986
   ],
   "1_1_async_actor_calls_async":[
      3600.07216861102,
      21.572291078217596
   ],
   "1_1_async_actor_calls_with_args_async":[
      2479.2791025030774,
      15.100892182939164
   ],
   "1_n_async_actor_calls_async":[
      6348.851031034499,
      61.198081662692225
   ],
   "n_n_async_actor_calls_async":[
      8828.83966538193,
      268.6117864610785
   ],
   "placement_group_create/removal":[
      1164.2409050338586,
      13.066684402747692
   ],
   "client__get_calls":[
      1536.189691591998,
      19.331772082393528
   ],
   "client__put_calls":[
      721.6739389457572,
      0.24234565062561444
   ],
   "client__put_gigabytes":[
      0.046168727223296846,
      6.763820926335119e-05
   ],
   "client__tasks_and_put_batch":[
      9011.064052826105,
      74.29753576149525
   ],
   "client__1_1_actor_calls_sync":[
      563.7096081091545,
      0.7244106759641209
   ],
   "client__1_1_actor_calls_async":[
      845.8681489979724,
      6.513902706128811
   ],
   "client__1_1_actor_calls_concurrent":[
      846.4716110735912,
      2.677132455876885
   ],
   "client__tasks_and_get_batch":[
      0.8017921868970177,
      0.011418693394455696
   ]
}
```

versus this PR
```
      2443.171686426444,
      7.596210165488628
   ],
   "1_n_async_actor_calls_async":[
      6075.208224257781,
      18.0449636629851
   ],
   "n_n_async_actor_calls_async":[
      8466.515151222513,
      183.94719482435514
   ],
   "placement_group_create/removal":[
      1101.2267618304986,
      7.213535574857593
   ],
   "client__get_calls":[
      1548.2243438966946,
      31.920527489054535
   ],
   "client__put_calls":[
      716.2383036529125,
      0.9816321516008939
   ],
   "client__put_gigabytes":[
      0.04608069565429085,
      0.0001254142930314185
   ],
   "client__tasks_and_put_batch":[
      8796.644395488453,
      45.830670773349766
   ],
   "client__1_1_actor_calls_sync":[
      563.2689170739174,
      0.4624398546907118
   ],
   "client__1_1_actor_calls_async":[
      845.527825397318,
      7.992744530415372
   ],
   "client__1_1_actor_calls_concurrent":[
      844.4180043946874,
      6.653925671888334
   ],
   "client__tasks_and_get_batch":[
      0.7994957585678762,
      0.012720877709905945
   ]
}
```
the per change seems not significant

**Pull request merged**